### PR TITLE
feat: Parse tomes as gears, show tome identification percentages

### PR DESCRIPTION
**Pull request by kristofbolyai**
This PR is best reviewed per commit, to understand the underlying refactors needed for it to work. Some commits have descriptions, to help understanding the commits. The size of the PR is not that huge as it seems, especially when going through commits 1 by 1.

Tomes are really close to actual gear items, but they need some adaptations, hence the refactors. In the end, the new system behaves in a much more adaptive way, allowing us to implement new item types effectively.

Some screenshots:
image
image
image
image
image
image

### CODE REVIEW

File: common/src/main/java/com/wynntils/models/gear/GearInfoRegistry.java [hunk 1]
```diff
@@ -89,7 +68,7 @@ private void loadGearRegistry() {
         Download dl = Managers.Net.download(UrlId.DATA_STATIC_GEAR_ADVANCED);
         dl.handleJsonObject(json -> {
             Gson gson = new GsonBuilder()
-                    .registerTypeHierarchyAdapter(GearInfo.class, new GearInfoDeserializer())
+                    .registerTypeHierarchyAdapter(GearInfo.class, new GearInfoDeserizalier())
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/gear/GearInfoRegistry.java]: typo in the name

File: common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.models.gear.type;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java]: Not sure the type package is the right place for this. We usually reserve it for interfaces, enums and records.

File: common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+        }
+
+        return List.copyOf(obtainInfo);
+    }
+
+    protected Optional<StyledText> parseLore(JsonObject json) {
+        String lore = JsonUtils.getNullableJsonString(json, "lore");
+        if (lore == null) return Optional.empty();
+
+        // Some lore contain like "\\[Community Event Winner\\]", fix that
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java]: Just noticed this: Is this still relevant in the v3 api?

File: common/src/main/java/com/wynntils/models/tooltip/IdentifiableItemInfo.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.models.tooltip;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/tooltip/IdentifiableItemInfo.java]: move to `type` subpackage.

File: common/src/main/java/com/wynntils/models/tooltip/TooltipIdentificationDecorator.java [hunk 5]
```diff
@@ -2,14 +2,13 @@
  * Copyright © Wynntils 2023.
  * This file is released under LGPLv3. See LICENSE for full license details.
  */
-package com.wynntils.models.gear.tooltip;
+package com.wynntils.models.tooltip;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/tooltip/TooltipIdentificationDecorator.java]: Should also be in `type`, I think.

File: common/src/main/java/com/wynntils/models/tooltip/TooltipStyle.java [hunk 6]
```diff
@@ -2,11 +2,11 @@
  * Copyright © Wynntils 2023.
  * This file is released under LGPLv3. See LICENSE for full license details.
  */
-package com.wynntils.models.gear.tooltip;
+package com.wynntils.models.tooltip;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/models/tooltip/TooltipStyle.java]: `type`

File: common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java (hunk 2 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java]: I think it better belongs in `models.wynnitem`, an "abstract" model I tried to create to unify similar item kinds. I think this has the same kind of vibe to it.

**Comment by magicus** [comment]
Call me crazy, but I'm seriously thinking that the new `tooltip` model would be better of as a `handler` than a model. According to the javadoc guidelines:

```
 * Models are our representation of the Wynncraft world, game state and Wynncraft
 * metadata. In general, any non-trivial interesting aspect of Wynncraft that is
 * provided through vanilla Minecraft elements, should be parsed and provided by
 * a Model.
```

and

```
 * Handlers span the bridge between Minecraft and Wynncraft. They manage a certain
 * aspect of Minecraft functionality, and with some Wynncraft knowledge, they distribute
 * the MC content to the actual Wynncraft models. A handler is needed when there is not a
 * clear 1-to-1 relationship between Minecraft components and models.
```

In this case it is not really "dispatching", but I think it is the same core idea. We need to provide a service to several models to be able to build tooltips. That is not so much a model per se, as a support layer for the models.

**Comment by magicus** [comment]
Overall, I think this look pretty nifty!

File: common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java (hunk 3 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java]: Removed it, it's not.

**Comment by kristofbolyai** [comment]
I also moved the TooltipModel to be a TooltipHandler. IMO it feels just as weird, but I can see how this is theoretically a Handler, it's just that we haven't made any of this kind.

**Review by magicus — APPROVED** [review]

**Pull request merged**